File: src/main/java/edu/kruchon/natural/lang/parser/freemarker/TemplateProcessor.java
package edu.kruchon.natural.lang.parser.freemarker;

import edu.kruchon.natural.lang.parser.exceptions.NaturalLangParserUncheckedException;
import freemarker.template.Configuration;
import freemarker.template.Template;
import freemarker.template.TemplateException;
import freemarker.template.TemplateExceptionHandler;

import java.io.File;
import java.io.IOException;
import java.io.StringWriter;
import java.io.Writer;
import java.util.Map;
import java.util.Objects;

public enum TemplateProcessor {
    INSTANCE;

    private Configuration cfg;

    TemplateProcessor() {
        try {
            cfg = new Configuration(Configuration.VERSION_2_3_20);
            cfg.setDirectoryForTemplateLoading(new File(Objects.requireNonNull(this.getClass().getClassLoader().getResource("./templates")).getFile()));
            cfg.setDefaultEncoding("UTF-8");
            cfg.setTemplateExceptionHandler(TemplateExceptionHandler.RETHROW_HANDLER);
            cfg.setLogTemplateExceptions(false);
            cfg.setWrapUncheckedExceptions(true);
            cfg.setFallbackOnNullLoopVariable(false);
        } catch (IOException e) {
            throw new NaturalLangParserUncheckedException(e);
        }
    }

    public String process(String templateName, Map<String, Object> templateParameters) {
        try (Writer out = new StringWriter()) {
            Template template = cfg.getTemplate(templateName);
            template.process(templateParameters, out);
            return out.toString();
        } catch (IOException | TemplateException e) {
            throw new NaturalLangParserUncheckedException("Cannot process template " + templateName, e);
        }
    }
}

File: src/main/kotlin/edu/kruchon/natural/lang/parser/freemarker/TemplateProcessor.kt
package edu.kruchon.natural.lang.parser.freemarker

import edu.kruchon.natural.lang.parser.exceptions.NaturalLangParserUncheckedException
import edu.kruchon.natural.lang.parser.generation.KotlinGenerationProperties
import edu.kruchon.natural.lang.parser.generation.KotlinGenerationUtils
import freemarker.template.Configuration
import freemarker.template.TemplateException
import freemarker.template.TemplateExceptionHandler
import java.io.File
import java.io.IOException
import java.io.StringWriter

internal object TemplateProcessor {

    private val cfg: Configuration = Configuration(Configuration.VERSION_2_3_20).apply {
        try {
            this.setDirectoryForTemplateLoading(File(checkNotNull(this.javaClass.classLoader.getResource("./templates")).file))
            this.defaultEncoding = "UTF-8"
            this.templateExceptionHandler = TemplateExceptionHandler.RETHROW_HANDLER
            this.logTemplateExceptions = false
            this.wrapUncheckedExceptions = true
            this.fallbackOnNullLoopVariable = false
        } catch (e: IOException) {
            throw NaturalLangParserUncheckedException(e)
        }
    }

    fun process(templateName: String, templateParameters: Map<String, Any>): String {
        try {
            StringWriter().use { out ->
                val implementationPackage = KotlinGenerationProperties.implementationPackage
                    ?: throw NaturalLangParserUncheckedException("Implementation package is not defined in object KotlinGenerationProperties")
                val commonTemplateParameters = mapOf(
                    "generationPackage" to KotlinGenerationProperties.generationPackage,
                    "implementationPackage" to implementationPackage
                )
                val template = cfg.getTemplate(templateName)
                template.process(
                    templateParameters + commonTemplateParameters, out
                )
                return out.toString()
            }
        } catch (e: IOException) {
            throw NaturalLangParserUncheckedException("Cannot process template $templateName", e)
        } catch (e: TemplateException) {
            throw NaturalLangParserUncheckedException("Cannot process template $templateName", e)
        }
    }
}